README.md
<h1 align="center">hyper-adapter-fs</h1>
<p align="center">A Storage port adapter that uses the local file system for object storage in the <a href="https://hyper.io/">hyper</a>  service framework</p>
</p>
<p align="center">
  <a href="https://nest.land/package/hyper-adapter-fs"><img src="https://nest.land/badge.svg" alt="Nest Badge" /></a>
  <a href="https://github.com/hyper63/hyper-adapter-fs/actions/workflows/test-and-publish.yml"><img src="https://github.com/hyper63/hyper-adapter-fs/actions/workflows/test-and-publish.yml/badge.svg" alt="Test" /></a>
  <a href="https://github.com/hyper63/hyper-adapter-fs/tags/"><img src="https://img.shields.io/github/tag/hyper63/hyper-adapter-fs" alt="Current Version" /></a>
</p>

---

## Table of Contents

- [Getting Started](#getting-started)
- [Installation](#installation)
- [Features](#features)
- [Methods](#methods)
- [Contributing](#contributing)
- [License](#license)

## Getting Started

`hyper.config.js`

```js
import { default as fs } from 'https://x.nest.land/hyper-adapter-fs@2.0.3/mod.js'

export default {
  app,
  adapter: [{ port: 'storage', plugins: [fs({ dir: './data' })] }],
}
```

## Installation

This is a Deno module available to import from
[nest.land](https://nest.land/package/hyper-adapter-fs)

deps.js

```js
export { default as fs } from 'https://x.nest.land/hyper-adapter-fs@2.0.3/mod.js'
```

## Features

- Create an bucket as a directory
- Remove an bucket as a directory
- List buckets as directories
- Put an object into a bucket as a directory
- Remove an object from a bucket as a directory
- Get an object from a bucket as a directory
- List objects in a bucket as a directory

## Methods

This adapter fully implements the Storage port and can be used as the
[hyper Storage service](https://docs.hyper.io/storage-api) adapter

See the full port [here](https://nest.land/package/hyper-port-storage)

## Contributing

Contributions are welcome! See the hyper
[contribution guide](https://docs.hyper.io/contributing-to-hyper)

## Testing

```
deno task test
```

To lint, check formatting, and run unit tests

## License

Apache-2.0


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [commit-and-tag-version](https://github.com/absolute-version/commit-and-tag-version) for commit guidelines.

## [3.0.2](https://github.com/hyper63/hyper-adapter-fs/compare/v3.0.1...v3.0.2) (2023-11-28)

## [3.0.1](https://github.com/hyper63/hyper-adapter-fs/compare/v3.0.0...v3.0.1) (2023-05-23)

## [3.0.0](https://github.com/hyper63/hyper-adapter-fs/compare/v2.1.0...v3.0.0) (2023-05-17)


### ⚠ BREAKING CHANGES

* putObject accepts a ReadableStream and getObject
returns a ReadableStream.

### Features

* beef up coverage and address issues [#10](https://github.com/hyper63/hyper-adapter-fs/issues/10) ([06b3e16](https://github.com/hyper63/hyper-adapter-fs/commit/06b3e16ff33770491db53d429e08d544e6b3b129))
* getObject and putObject return and accept a ReadableStream [#10](https://github.com/hyper63/hyper-adapter-fs/issues/10) ([a16e53d](https://github.com/hyper63/hyper-adapter-fs/commit/a16e53d34397248cbbd40225b881f5919ae4e516))

## [2.1.0](https://github.com/hyper63/hyper-adapter-fs/compare/v2.0.3...v2.1.0) (2022-08-30)


### Features

* useSignedUrl returns 501 for getObject and putObject [#7](https://github.com/hyper63/hyper-adapter-fs/issues/7) [#8](https://github.com/hyper63/hyper-adapter-fs/issues/8) ([c3f4ece](https://github.com/hyper63/hyper-adapter-fs/commit/c3f4ece16684e74dc5a067d6a08b6b077c62b734))

### [2.0.3](https://github.com/hyper63/hyper-adapter-fs/compare/v2.0.2...v2.0.3) (2022-04-20)


### Bug Fixes

* 404 if bucket not found [#5](https://github.com/hyper63/hyper-adapter-fs/issues/5) ([68c7c10](https://github.com/hyper63/hyper-adapter-fs/commit/68c7c105f6751a97943685b8b4c854e53dacbfa7))
* 409 on makeBucket if already exists ([5561537](https://github.com/hyper63/hyper-adapter-fs/commit/55615377494e56ecc386764a3ed0f7fca6e97141))


test/hyper.js
// Harness deps
import { default as appOpine } from "https://x.nest.land/hyper-app-opine@2.2.0/mod.js";
import { default as core } from "https://x.nest.land/hyper@3.3.0/mod.js";

import fs from "../mod.js";

const hyperConfig = {
  app: appOpine,
  adapters: [
    { port: "storage", plugins: [fs({ dir: "./tmp" })] },
  ],
};

core(hyperConfig);


adapter.test.js
import { assert, assertEquals, assertObjectMatch } from './dev_deps.js'

import createAdapter from './adapter.js'

const v4 = () => crypto.randomUUID()

const adapter = createAdapter('./tmp')

for await (const dirEntry of Deno.readDir('./tmp')) {
  if (dirEntry.isDirectory) {
    console.log('deleting test bucket: ', dirEntry)
    await Deno.remove(`./tmp/${dirEntry.name}`, { recursive: true })
  }
}

Deno.test('adapter', async (t) => {
  await t.step('all', async (t) => {
    await t.step(
      'should return a HyperErr with status 404 if the folder is not found',
      async () => {
        const object = v4() + '.txt'

        // test
        const stream = new Response(new TextEncoder().encode('woop woop')).body

        const result = await adapter.putObject({ bucket: 'dne', object, stream })

        assert(!result.ok)
        assertEquals(result.status, 404)
      },
    )

    await t.step(
      'should return a HyperErr if the bucket name is contains relative parts',
      async () => {
        const bucket = v4()
        const err = await adapter.makeBucket(`../${bucket}`)

        assertObjectMatch(err, {
          ok: false,
          status: 400,
          msg: 'cannot contain relative path parts',
        })
      },
    )

    await t.step(
      'should return a HyperErr with status 400 if the bucket name contains slashes',
      async () => {
        const bucket = v4()
        const err = await adapter.makeBucket(`${bucket}/foo`)

        assertObjectMatch(err, {
          ok: false,
          status: 400,
          msg: 'bucket name cannot contain slashes',
        })

        const backwardsSlash = await adapter.makeBucket(`${bucket}\\foo`)

        assertObjectMatch(backwardsSlash, {
          ok: false,
          status: 400,
          msg: 'bucket name cannot contain slashes',
        })
      },
    )
  })

  await t.step('makeBucket', async (t) => {
    await t.step('should create the folder', async () => {
      const bucket = v4()
      const result = await adapter.makeBucket(bucket)
      assert(result.ok)
      await adapter.removeBucket(bucket)
    })

    await t.step(
      'should return a HyperErr with status 409 if the bucket already exists',
      async () => {
        const bucket = v4()
        await adapter.makeBucket(bucket)
        const result = await adapter.makeBucket(bucket)

        assert(!result.ok)
        assertEquals(result.status, 409)
        await adapter.removeBucket(bucket)
      },
    )
  })

  await t.step('removeBucket', async (t) => {
    await t.step('should remove the folder', async () => {
      const bucket = v4()
      await adapter.makeBucket(bucket)
      const result = await adapter.removeBucket(bucket)
      assert(result.ok)
    })

    await t.step(
      'should return a HyperErr with status 404 if the folder does not exist',
      async () => {
        const bucket = v4()
        const result = await adapter.removeBucket(bucket)
        assertObjectMatch(result, {
          ok: false,
          status: 404,
          msg: 'bucket does not exist',
        })
      },
    )
  })

  await t.step('putObject', async (t) => {
    await t.step('should put the object into the folder', async () => {
      // setup
      const bucket = v4()
      const object = v4() + '.txt'
      await adapter.makeBucket(bucket)

      // test
      const stream = new Response('woop woop').body

      const result = await adapter.putObject({ bucket, object, stream })
      assert(result.ok)

      // clean up

      // remove file
      await adapter.removeObject({ bucket, object })
      // remove Bucket
      await adapter.removeBucket(bucket)
    })

    await t.step('should put the object into the sub-folder', async () => {
      // setup
      const bucket = v4()
      const object = `foo/${v4() + '.txt'}`
      await adapter.makeBucket(bucket)

      // test
      const stream = new Response('woop woop').body

      const result = await adapter.putObject({ bucket, object, stream })
      assert(result.ok)

      // clean up

      // remove file
      await adapter.removeObject({ bucket, object })
      // remove Bucket
      await adapter.removeBucket(bucket)
    })

    await t.step(
      'should return a HyperErr with a status of 501 when requesting a signed url',
      async () => {
        // setup
        const bucket = v4()
        const object = v4() + '.txt'
        await adapter.makeBucket(bucket)

        // test
        const result = await adapter.putObject({ bucket, object, useSignedUrl: true }).catch((
          err,
        ) => err)

        assertObjectMatch(result, {
          ok: false,
          status: 501,
        })
        // remove Bucket
        await adapter.removeBucket(bucket)
      },
    )
  })

  await t.step('getObject', async (t) => {
    await t.step('should retrieve the object from the folder', async () => {
      const bucket = v4()
      const object = v4() + '.txt'
      await adapter.makeBucket(bucket)

      const stream = new Response('hello world').body

      await adapter.putObject({
        bucket,
        object,
        stream,
      })
      // test
      const s = await adapter.getObject({ bucket, object })

      const result = await new Response(s).text()

      assertEquals(result, 'hello world')

      // cleanup
      // remove file
      await adapter.removeObject({ bucket, object })
      // remove Bucket
      await adapter.removeBucket(bucket)
    })

    await t.step(
      'should return a HyperErr with status 404 if the object does not exist',
      async () => {
        const bucket = v4()
        const object = v4() + '.txt'
        await adapter.makeBucket(bucket)

        // test
        const result = await adapter.getObject({ bucket, object })

        assertObjectMatch(result, {
          ok: false,
          status: 404,
          msg: 'object does not exist',
        })
        // remove Bucket
        await adapter.removeBucket(bucket)
      },
    )

    await t.step(
      'should return a HyperErr with a status of 501 when requesting a signed url',
      async () => {
        // setup
        const bucket = v4()
        const object = v4() + '.txt'
        await adapter.makeBucket(bucket)

        // test
        const result = await adapter.getObject({ bucket, object, useSignedUrl: true }).catch((
          err,
        ) => err)
        const getResult = await adapter.getObject({ bucket, object })

        assertObjectMatch(getResult, {
          ok: false,
          status: 404,
          msg: 'object does not exist',
        })

        assertObjectMatch(result, {
          ok: false,
          status: 501,
        })
        // remove Bucket
        await adapter.removeBucket(bucket)
      },
    )
  })

  await t.step('removeObject', async (t) => {
    await t.step('should remove the object', async () => {
      const bucket = v4()
      const object = v4() + '.txt'
      await adapter.makeBucket(bucket)

      const stream = new Response('hello world').body

      await adapter.putObject({
        bucket,
        object,
        stream,
      })
      // test
      const res = await adapter.removeObject({ bucket, object })
      assert(res.ok)

      // cleanup
      // remove Bucket
      await adapter.removeBucket(bucket)
    })

    await t.step(
      'should return a HyperErr with status 404 if the object does not exist',
      async () => {
        const bucket = v4()
        const object = v4() + '.txt'
        await adapter.makeBucket(bucket)

        // test
        const result = await adapter.removeObject({ bucket, object })

        assertObjectMatch(result, {
          ok: false,
          status: 404,
          msg: 'object does not exist',
        })
        // remove Bucket
        await adapter.removeBucket(bucket)
      },
    )
  })

  await t.step('listObjects', async (t) => {
    await t.step('should list the objects in the path', async () => {
      const bucket = v4()
      const object = v4() + '.tmp'

      // setup
      await adapter.makeBucket(bucket)
      const stream = new Response('woop woop').body
      await adapter.putObject({ bucket, object, stream })

      // test
      const { objects } = await adapter.listObjects({ bucket })

      assert(objects.find((file) => file === object))

      // clean up
      await adapter.removeObject({ bucket, object })
      await adapter.removeBucket(bucket)
    })

    await t.step(
      'should return an empty list of objects if the prefix does not exist',
      async () => {
        const bucket = v4()
        const object = v4() + '.tmp'

        // setup
        await adapter.makeBucket(bucket)
        const stream = new Response('woop woop').body
        await adapter.putObject({ bucket, object, stream })

        // test
        const { objects } = await adapter.listObjects({ bucket, prefix: '/foo' })

        assertEquals(objects.length, 0)
        await adapter.removeBucket(bucket)
      },
    )

    await t.step(
      'should return a HyperErr with status 404 if the bucket does not exist',
      async () => {
        const bucket = v4()

        // test
        const result = await adapter.listObjects({ bucket })

        assertObjectMatch(result, {
          ok: false,
          status: 404,
          msg: 'bucket does not exist',
        })
      },
    )
  })
})


adapter_test.js
import {
  assert,
  assertEquals,
  assertObjectMatch,
  readAll,
  v4 as v4Generator,
} from "./dev_deps.js";

import createAdapter from "./adapter.js";

const v4 = v4Generator.generate.bind(v4Generator);

const adapter = createAdapter("./tmp");

function emptyReader() {
  return {
    read(_) {
      return Promise.resolve(null);
    },
  };
}

/**
 * Given a string, return a Reader that read the encoded string
 * into the provided buffer
 *
 * @param {string} text - the string to stream
 * @returns {Deno.Reader} - a Reader implementation
 */
function textReader(text = "") {
  const encoded = new TextEncoder().encode(text);
  let totalRead = 0;
  let finished = false;

  async function read(buf) {
    if (finished) {
      return null;
    }

    let result;
    const remaining = encoded.length - totalRead;

    // read into the buffer
    buf.set(encoded.subarray(totalRead, buf.byteLength), 0);

    if (remaining >= buf.byteLength) {
      result = buf.byteLength;
    } else {
      result = remaining;
    }

    if (result) {
      totalRead += result;
    }
    finished = totalRead === encoded.length;

    return await result;
  }

  return { read };
}

for await (const dirEntry of Deno.readDir("./tmp")) {
  if (dirEntry.isDirectory) {
    console.log("deleting test bucket: ", dirEntry);
    await Deno.remove(`./tmp/${dirEntry.name}`, { recursive: true });
  }
}

Deno.test("fs adapter make bucket", async () => {
  const bucket = v4();
  const result = await adapter.makeBucket(bucket);
  assert(result.ok);
  await adapter.removeBucket(bucket);
});

Deno.test("fs adapter make bucket - invalid name", async () => {
  const bucket = v4();
  const err = await adapter.makeBucket(`../${bucket}`);

  assertObjectMatch(err, {
    ok: false,
    status: 400,
    msg: "bucket name cannot contain relative path parts",
  });
});

Deno.test("fs adapter make bucket - 409 if already exists", async () => {
  const bucket = v4();
  await adapter.makeBucket(bucket);
  const result = await adapter.makeBucket(bucket);

  console.log(result);

  assert(!result.ok);
  assertEquals(result.status, 409);
  await adapter.removeBucket(bucket);
});

Deno.test("fs adapter remove bucket", async () => {
  const bucket = v4();
  await adapter.makeBucket(bucket);
  const result = await adapter.removeBucket(bucket);
  assert(result.ok);
});

Deno.test("fs adapter remove bucket - 404 if not found", async () => {
  const bucket = v4();
  const result = await adapter.removeBucket(bucket);
  assert(!result.ok);
  assertEquals(result.status, 404);
});

Deno.test("fs adapter all methods 404 - if bucket not found", async () => {
  const object = v4() + ".txt";

  // test
  const stream = textReader("woop woop");

  const result = await adapter.putObject({
    bucket: "dne",
    object,
    stream,
  });

  assert(!result.ok);
  assertEquals(result.status, 404);
});

Deno.test("fs adapter put object", async () => {
  // setup
  const bucket = v4();
  const object = v4() + ".txt";
  await adapter.makeBucket(bucket);

  // test
  const stream = textReader("woop woop");

  const result = await adapter.putObject({
    bucket,
    object,
    stream,
  });
  assert(result.ok);

  // clean up

  // remove file
  await adapter.removeObject({
    bucket,
    object,
  });
  // remove Bucket
  await adapter.removeBucket(bucket).catch((err) => {
    console.log(JSON.stringify(err));
    return { ok: false };
  });
});

Deno.test("fs adapter put object - useSignedUrl", async () => {
  // setup
  const bucket = v4();
  const object = v4() + ".txt";
  await adapter.makeBucket(bucket);

  // test
  const result = await adapter.putObject({
    bucket,
    object,
    useSignedUrl: true,
  }).catch((err) => err);
  assert(!result.ok);
  assert(result.status === 501);
});

Deno.test("fs adapter get object", async () => {
  const bucket = v4();
  const object = v4() + ".txt";
  await adapter.makeBucket(bucket);

  const stream = textReader("hello world");

  await adapter.putObject({
    bucket,
    object,
    stream,
  });
  // test
  const s = await adapter.getObject({
    bucket,
    object,
  });

  const encodedResult = await readAll(s);
  // close the Reader
  s.close();

  assertEquals(new TextDecoder().decode(encodedResult), "hello world");

  // cleanup
  // remove file
  await adapter.removeObject({
    bucket,
    object,
  });
  // remove Bucket
  await adapter.removeBucket(bucket).catch(() => {
    return { ok: false };
  });
});

Deno.test("fs adapter get object - useSignedUrl", async () => {
  // setup
  const bucket = v4();
  const object = v4() + ".txt";
  await adapter.makeBucket(bucket);

  // test
  const result = await adapter.getObject({
    bucket,
    object,
    useSignedUrl: true,
  }).catch((err) => err);
  assert(!result.ok);
  assert(result.status === 501);
});

Deno.test("list files", async () => {
  const bucket = v4();
  const object = v4() + ".tmp";

  // setup
  await adapter.makeBucket(bucket);

  const stream = emptyReader();

  await adapter.putObject({
    bucket,
    object,
    stream,
  });

  // test
  const list = await adapter.listObjects({
    bucket,
  });

  assert(
    list.find((file) => file === object),
  );

  // clean up
  await adapter.removeObject({
    bucket,
    object,
  });
  await adapter.removeBucket(bucket);
});


adapter.js
import { crocks, HyperErr, isHyperErr, path, R } from './deps.js'

const { Async } = crocks
const { always, is, ifElse } = R

export const handleHyperErr = ifElse(
  isHyperErr,
  Async.Resolved,
  Async.Rejected,
)

/**
 * hyper63 adapter for the storage port
 *
 * This adapter uses the file system
 * as the implementation details for
 * the storage port.
 *
 * @typedef {Object} StorageInfo
 * @property {string} bucket
 * @property {string} object
 *
 * @typedef {Object} StorageObject
 * @property {string} bucket
 * @property {string} object
 * @property {stream} stream
 *
 * @typedef {Object} Response
 * @property {boolean} ok
 * @property {string} [msg] - error message
 */
/**
 * @param {string} path
 * @returns {Object}
 */
export default function (root) {
  if (!root) {
    throw new Error(
      'STORAGE: FS_Adapter: root directory required for this service!',
    )
  }

  const open = Async.fromPromise(Deno.open.bind(Deno))
  const mkdir = Async.fromPromise(Deno.mkdir.bind(Deno))
  const rm = Async.fromPromise(Deno.remove.bind(Deno))
  const rmdir = Async.fromPromise(Deno.remove.bind(Deno))
  const stat = Async.fromPromise(Deno.stat.bind(Deno))

  const resolvePathFromRoot = (...pieces) => path.resolve(path.join(root, ...pieces))

  const checkRelativeParts = (path) =>
    Async.of(path.includes('..'))
      .chain((invalid) =>
        !invalid ? Async.Resolved(path) : Async.Rejected(
          HyperErr({
            status: 400,
            msg: 'cannot contain relative path parts',
          }),
        )
      )

  const checkBucketName = (name) =>
    checkRelativeParts(name)
      /**
       * A bucket name also cannot contain slashes
       */
      .map((name) => name.includes('/') || name.includes('\\'))
      .chain((invalid) =>
        !invalid ? Async.Resolved(name) : Async.Rejected(
          HyperErr({
            status: 400,
            msg: 'bucket name cannot contain slashes',
          }),
        )
      )

  const checkPathExists = (path, resource = 'bucket') =>
    Async.of(path)
      .chain(stat)
      .bimap(
        (err) => {
          if (is(Deno.errors.NotFound)) {
            return HyperErr({ status: 404, msg: `${resource} does not exist` })
          }
          return err
        },
        always(path),
      )

  const checkObjectPath = ({ bucket, object }) =>
    Async.all([
      // Bucket checks
      Async.all([
        checkBucketName(bucket),
        checkPathExists(resolvePathFromRoot(bucket)),
      ]),
      // Object checks
      Async.all([
        checkRelativeParts(object),
        checkPathExists(resolvePathFromRoot(bucket, object), 'object'),
      ]),
    ])
      .map(() => resolvePathFromRoot(bucket, object))

  /**
   * @param {string} name
   * @returns {Promise<Response>}
   */
  function makeBucket(name) {
    return Async.of(name)
      .chain(checkBucketName)
      .map((name) => resolvePathFromRoot(name))
      .chain((path) =>
        checkPathExists(path).bichain(
          () => Async.Resolved(path),
          () =>
            Async.Rejected(
              HyperErr({ status: 409, msg: 'bucket already exists' }),
            ),
        )
      )
      // Make the directory as the new bucket
      .chain((dir) => mkdir(dir, { recursive: true }))
      .map(always({ ok: true }))
      .bichain(
        handleHyperErr,
        Async.Resolved,
      )
      .toPromise()
  }

  /**
   * @param {string} name
   * @returns {Promise<Response>}
   */
  function removeBucket(name) {
    return Async.of(name)
      .chain(checkBucketName)
      .chain((name) => checkPathExists(resolvePathFromRoot(name)))
      // Delete all contents of directory, then directory
      .chain((bucket) => rmdir(bucket, { recursive: true }))
      .map(always({ ok: true }))
      .bichain(
        handleHyperErr,
        Async.Resolved,
      )
      .toPromise()
  }

  /**
   * @param {StorageObject}
   * @returns {Promise<Response>}
   */
  function putObject({ bucket, object, stream, useSignedUrl }) {
    if (useSignedUrl) {
      return Promise.resolve(
        HyperErr({
          status: 501,
          msg: 'Not Implemented',
        }),
      )
    }

    return Async.all([
      // Bucket checks
      Async.all([
        checkBucketName(bucket),
        checkPathExists(resolvePathFromRoot(bucket)),
      ]),
      // Object checks
      Async.all([
        checkRelativeParts(object),
      ]),
    ])
      .map(() => resolvePathFromRoot(bucket, object))
      /**
       * Check if the sub-directory in the bucket exists,
       * and create it if it does not
       */
      .chain((p) =>
        checkPathExists(path.dirname(p))
          .bichain(
            /**
             * The directory does not exist within the bucket, so create it
             */
            () => mkdir(path.dirname(p), { recursive: true }).map(() => p),
            /**
             * The directory does exist with the bucket, so simply noop
             */
            () => Async.Resolved(p),
          )
      )
      .chain((path) => {
        return open(path, { create: true, write: true, truncate: true })
      })
      .chain(Async.fromPromise((file) => {
        return stream.pipeTo(file.writable)
      }))
      .map(always({ ok: true }))
      .bichain(
        handleHyperErr,
        Async.Resolved,
      ).toPromise()
  }

  /**
   * @param {StorageInfo}
   * @returns {Promise<Response>}
   */
  function removeObject({ bucket, object }) {
    return checkObjectPath({ bucket, object })
      .map(() => resolvePathFromRoot(bucket, object))
      .chain(rm)
      .map(always({ ok: true }))
      .bichain(
        handleHyperErr,
        Async.Resolved,
      ).toPromise()
  }

  /**
   * @param {StorageInfo}
   * @returns {Promise<stream>}
   */
  function getObject({ bucket, object, useSignedUrl }) {
    if (useSignedUrl) {
      return Promise.resolve(
        HyperErr({
          status: 501,
          msg: 'Not Implemented',
        }),
      )
    }

    return checkObjectPath({ bucket, object })
      .chain((p) => open(p, { read: true, write: false }))
      .map((file) => file.readable)
      .bichain(
        handleHyperErr,
        Async.Resolved,
      ).toPromise()
  }

  function listObjects({ bucket, prefix = '' }) {
    return Async.all([
      checkBucketName(bucket),
      checkPathExists(resolvePathFromRoot(bucket)),
    ])
      .chain(() =>
        /**
         * At this point, we know the bucket exists, but the prefix may not,
         * but that shouldn't result in an error
         *
         * So if the prefix does not exist we just return an empty list of objects
         */
        checkPathExists(resolvePathFromRoot(bucket, prefix)).bichain(
          () => Async.Resolved({ ok: true, objects: [] }),
          Async.fromPromise(async (path) => {
            const files = []
            try {
              for await (const dirEntry of Deno.readDir(path)) {
                files.push(dirEntry.name)
              }

              return { ok: true, objects: files }
            } catch (err) {
              if (is(Deno.errors.NotFound)) {
                return HyperErr({ status: 404, msg: `${resource} does not exist` })
              }
              throw err
            }
          }),
        )
      )
      .bichain(
        handleHyperErr,
        Async.Resolved,
      ).toPromise()
  }

  return Object.freeze({
    makeBucket,
    removeBucket,
    listBuckets: () => Promise.resolve(null),
    putObject,
    removeObject,
    getObject,
    listObjects,
  })
}
